### PR TITLE
[ALLUXIO-2441] Streaming data transport

### PR DESCRIPTION
**Pull request by peisun1115**
https://alluxio.atlassian.net/browse/ALLUXIO-2441

Merge streaming branch to master

See more description of this change in https://github.com/Alluxio/alluxio/pull/4427 and https://alluxio.atlassian.net/browse/ALLUXIO-2441

### CODE REVIEW

File: core/client/pom.xml [hunk 1]
```diff
@@ -39,7 +39,7 @@
     <dependency>
       <groupId>io.netty</groupId>
       <artifactId>netty-all</artifactId>
-      <version>4.0.28.Final</version>
```
Comment (jsimsa) [inline core/client/pom.xml]: can we use dependency management for this?

File: core/client/src/main/java/alluxio/client/block/StreamFactory.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+ */
+@NotThreadSafe
+public final class StreamFactory {
+  private static final boolean PACKET_STREAMING_ENABLED =
+      Configuration.getBoolean(PropertyKey.USER_PACKET_STREAMING_ENABLED);
+
+  /**
+   * Private constructor.
+   */
+  private StreamFactory() {
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/block/StreamFactory.java]: the recommended style for this is:

```
private StreamFactory() {} // prevent instantiation
```

File: core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+   * @param options the options
+   * @throws IOException if an I/O error occurs
+   * @return the {@link BlockOutStream} instance created
+   */
+  public static BlockOutStream createLocalBlockOutStream(long blockId,
+      long blockSize,
+      WorkerNetAddress workerNetAddress,
+      FileSystemContext context,
+      OutStreamOptions options) throws IOException {
+    Closer closer = Closer.create();
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java]: can you try-with-resource and put the closer into the resource clause? if so, apply it throughout this PR

File: core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+   *
+   * @param blockId the block id
+   * @param blockSize the block size
+   * @param workerNetAddress the worker network address
+   * @param context the file system context
+   * @param options the options
+   * @throws IOException if an I/O error occurs
+   * @return the {@link BlockOutStream} instance created
+   */
+  public static BlockOutStream createLocalBlockOutStream(long blockId,
```
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java]: our convention is to conserver vertical space, but putting multiple arguments per line, as per IDE's formatting

Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java]: applies throughout this PR

File: core/common/pom.xml [hunk 5]
```diff
@@ -49,7 +49,7 @@
     <dependency>
       <groupId>io.netty</groupId>
       <artifactId>netty-all</artifactId>
-      <version>4.0.28.Final</version>
+      <version>4.1.6.Final</version>
```
Comment (jsimsa) [inline core/common/pom.xml]: use dependency management

File: core/common/src/main/java/alluxio/network/protocol/Status.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+   * @return true if the status is in OK state
+   */
+  public static boolean isOk(Protocol.Status status) {
+    return status.getCode() == Protocol.Status.Code.OK;
+  }
+
+  /**
+   * Private constructor.
+   */
+  private Status() {}
```
Comment (jsimsa) [inline core/common/src/main/java/alluxio/network/protocol/Status.java]: `private Status() {} // prevent instantiation`

File: core/server/pom.xml [hunk 7]
```diff
@@ -44,7 +44,7 @@
     <dependency>
       <groupId>io.netty</groupId>
       <artifactId>netty-all</artifactId>
-      <version>4.0.28.Final</version>
+      <version>4.1.6.Final</version>
```
Comment (jsimsa) [inline core/server/pom.xml]: dependency management

File: core/server/src/main/java/alluxio/worker/netty/NettyExecutors.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  public static final ExecutorService FILE_WRITER_EXECUTOR =
+      new ThreadPoolExecutor(THREADS_MIN,
+          Configuration.getInt(PropertyKey.WORKER_NETWORK_NETTY_FILE_WRITER_THREADS_MAX),
+          THREAD_STOP_MS, TimeUnit.MILLISECONDS, new SynchronousQueue<Runnable>(),
+          ThreadFactoryUtils.build("FilePacketWriterExecutor-%d", true));
+
+  /**
+   * Private constructor.
+   */
+  private NettyExecutors() {}
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/worker/netty/NettyExecutors.java]: `... // prevent instantiation`

File: core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java (hunk 3 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java]: It doesn't apply. They are closed only when exception is thrown.

**Comment by jsimsa** [comment]
@peisun1115 did a quick pass and left some comments

File: core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java (hunk 3 above)
Comment (jsimsa) [inline core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java]: ack

File: core/client/src/main/java/alluxio/client/block/StreamFactory.java (hunk 2 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/block/StreamFactory.java]: done

File: core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java (hunk 4 above)
Comment (peisun1115) [inline core/client/src/main/java/alluxio/client/block/stream/BlockOutStream.java]: done

File: core/common/src/main/java/alluxio/network/protocol/Status.java (hunk 6 above)
Comment (peisun1115) [inline core/common/src/main/java/alluxio/network/protocol/Status.java]: done

File: core/server/pom.xml (hunk 7 above)
Comment (peisun1115) [inline core/server/pom.xml]: done

File: core/common/pom.xml (hunk 5 above)
Comment (peisun1115) [inline core/common/pom.xml]: done

File: core/client/pom.xml (hunk 1 above)
Comment (peisun1115) [inline core/client/pom.xml]: done

**Comment by peisun1115** [comment]
Jenkins, test this please

**Comment by calvinjia** [comment]
LGTM, thanks again for implementing the first step in this important feature.

**Pull request merged**